manifests/server/config/user.pp

```diff
@@ -0,0 +1,120 @@
+# @summary Define ClickHouse users
+#
+# This type defines the ClickHouse users with specified parameters
+#
+# @param networks
+#   Defines network nodes where from user could connect
+#
+# @param profile
+#   User profile
+#
+# @param quota
+#   User quota
+#
+# @param user
+#   User name
+#
+# @param password
+#   Plaintext password. Will be hashed into sha256 format in xml file
+#
+# @param password_sha256
+#   Optional parameter. If defined, it will be used despite the value of `password`
+#
+# @param databases
+#   Optional hash of allowed DBs, tables and filters to implement ACLs
+#
+# @param allow_databases
+#   Optional parameter to restrict access to specified databases
+#
+# @param allow_dictionaries
+#   Optional parameter to restrict access to specified dictionaries
+#
+# @see https://clickhouse.tech/docs/en/operations/settings/settings_users/, https://clickhouse.tech/docs/v20.3/en/operations/access_rights/
+#
+# @example
+#   clickhouse::server::config::user { 'username':
+#       network   => {
+#           'ip'          => [
+#               '::',
+#               '0.0.0.0',
+#           ],
+#           'host'        => [
+#               'host1.local',
+#               'host2.local',
+#           ],
+#           'host_regexp' => [
+#               '[^.]*\.domain\.TLD',
+#           ],
+#       },
+#       profile   => 'profile_name',
+#       quota     => 'quota_name',
+#       password  => 'password',
+#       databases => {
+#           'db_name' => {
+#               'table_name'         => ['filter'],
+#               'another_table_name' => ['another filter'],
+#           },
+#       },
+#   }
+#
+# @author InnoGames GmbH
+#
+define clickhouse::server::config::user (
+    Struct[{
+        Optional[ip]          => Array[String[1], 1],
+        Optional[host]        => Array[String[1], 1],
+        Optional[host_regexp] => Array[String[1], 1],
+    }]                         $networks,
+    String[1]                  $profile            = 'default',
+    String[1]                  $quota              = 'default',
+    String[1]                  $user               = $title,
+    String[0]                  $password           = '',
+    Optional[
+        Pattern[/\A[0-9a-fA-F]{64}\Z/]
+    ]                          $password_sha256    = undef,
+    Optional[Hash]             $databases          = undef,
+    Optional[Array[String[1]]] $allow_databases    = undef,
```

Comment:
It's different APIs, and `allow_databases` is already deprecated, that's why the link to https://clickhouse.tech/docs/v20.3/en/operations/access_rights/ contains a specific version. Here `databases` are described https://clickhouse.tech/docs/en/operations/settings/settings-users/#user-namedatabases

I won't implement the new API here now, since it will be hard.